compiler/defer.go

```diff
@@ -127,12 +128,22 @@ func (c *Compiler) emitDefer(frame *Frame, instr *ssa.Defer) {
 		deferFrame = c.builder.CreateInsertValue(deferFrame, value, i, "")
 	}
 
-	// Put this struct in an alloca.
-	alloca := c.builder.CreateAlloca(deferFrameType, "defer.alloca")
-	c.builder.CreateStore(deferFrame, alloca)
-	if c.NeedsStackObjects() {
-		c.trackPointer(alloca)
+	// Put this struct in an allocation.
+	var alloca llvm.Value
+	if instr.Block().Dominates(instr.Parent().Blocks[0]) {
+		// This can safely use a stack allocation.
+		alloca = llvmutil.CreateEntryBlockAlloca(c.builder, deferFrameType, "defer.alloca")
+		if c.NeedsStackObjects() {
+			c.trackPointer(alloca)
+		}
+	} else {
+		// This may be hit a variable number of times, so use a heap allocation.
+		size := c.targetData.TypeAllocSize(deferFrameType)
+		sizeValue := llvm.ConstInt(c.uintptrType, size, false)
+		allocCall := c.createRuntimeCall("alloc", []llvm.Value{sizeValue}, "defer.alloc.call")
+		alloca = c.builder.CreateBitCast(allocCall, llvm.PointerType(deferFrameType, 0), "defer.alloc")
```

Comment:
This `alloca` should be tracked as well. So I would suggest moving the above lines:

```go

		if c.NeedsStackObjects() {
			c.trackPointer(alloca)
		}
```
to below this if/else.